### DISCUSSION TITLE
domoticz

### DISCUSSION THREAD

**Discussion by hotting** [ZuinigeRijder/SolisCloud2PVOutput#18] [Ideas]
As a user of domoticz i would like a version of the script that exports the data to domoticz is this possible?
Or does anyone have hints to do this i have limited knowledge of python

thank you for the effort taken
g Hans

**Comment by ZuinigeRijder** [comment]
Currently the script computes information and sends the data as comma separated list to PVOUTPUT:

````
            pvoutput_string = (
                'data=' + TODAY +  # Date
                ',' + datetime_current.strftime("%H:%M") +  # Time
                ',' + str(hi_res_watthour_today) +  # Energy Generation
                ',' + str(watt) +  # Power Generation
                ',-1' +  # no Energy Consumption
                ',' + str(ac_volt) +  # Power generation used for AC voltage
                ',' + str(inverter_temp) +  # inverter temp instead of outside temp
                ',' + str(volt)  # Voltage
            )
            send_pvoutput_data(pvoutput_string)
````

Instead or apart from sending pvoutput_string, you can also send this comma separated string to domoticz and further process it, like PVOUTPUT does. Note that the pvoutput_string is logged also to standard output, so you could also redirect this to domoticz? Especially, because the logging is now configurable.

I do not have experience with domoticz, so I do not know what is possible there.

**Comment by ZuinigeRijder** [comment]
Maybe this article will help you?

https://domoticx.com/internet-of-things-domoticz-data-ontvangen-vanuit-andere-bron/#

**Comment by hotting** [comment]
Thank you for the quick response.
In domoticz you can control devices via an API call. So I'm going to see how I can modify the pvout routine so that it can make an api call to domoticz
I will get to work with all the info, if I have a solution or have a question I will let you know

Thank you
Hans

**Comment by hotting** [comment]
After a few adjustments I managed to write the data to domoticz.
The devices in domoticz can be accessed via an API control. the syntacs for this can be found at
https://www.domoticz.com/wiki/Domoticz_API/JSON_URLs
Each value also needs its own device.

The adjustments I made are:

In the file "soliscloud_to_pvoutput.cfg" I created a section for domoticz with the ID number of the devices and the url where domoticz can be reached

```
[Domoticz]
domot_url = http://192.168.0.222:8081
domot_power_generated_id = 214
domot_ac_volt_id = 215
domot_inverter_temp_id = 0
domot_volt_id = 0
```


To be able to read this I have included in the script
```
# == read api_secrets in soliscloud_to_pvoutput.cfg ============================
made the following adjustment

Domoticz = dict(parser.items('Domoticz'))

# == domoticz devices, fill in yours in soliscloud_to_pvoutput.cfg ===========
DOMOT_URL = Domoticz['domot_url']
DOMOT_POWER_GENERATED_ID = Domoticz['domot_power_generated_id']
DOMOT_AC_VOLT_ID = Domoticz['domot_ac_volt_id']
DOMOT_INVERTER_TEMP_ID = Domoticz['domot_inverter_temp_id']
DOMOT_VOLT_ID = Domoticz['domot_volt_id']
```


Then created a routine that sends the value to domoticz in the style of the original script (read code used)

```
# == write to Domoticz ================================================= ===========
def send_to_Domoticz(IDX,Value):
     request = DOMOT_URL + "/json.htm?type=command&param=udevice&idx="+ IDX +"&svalue=" + Value
     error string = ''
     logging.debug(request)
     try:
         with urlopen(request, timeout=30) as response:
             body = response.read()
             content = body.decode("utf-8")
             logging.debug(content)
             return content
     except HTTPError as error:
         errorstring = str(error.status) + ': ' + error.reason
     except URLError as error:
         errorstring = str(error.reason)
     exceptTimeoutError:
         error string = 'Request timed out'
     except socket.timeout:
         errorstring = 'Socket timed out'
     except Exception as ex: # pylint: disable=broad-except
         errorstring = 'urlopen exception: ' + str(ex)
         traceback.print_exc()

     logging.error(request + " -> " + errorstring)
     time.sleep(60) # retry after 1 minute
     return 'ERROR'
```


And to put it to work the main routine completed with the following code.
In each line it is checked whether the device is active by checking whether the ID is greater than 0. If the device is active, the send_to_Domoticz routine is called


```
if DOMOT_POWER_GENERATED_ID : send_to_Domoticz(str(DOMOT_POWER_GENERATED_ID),str(watts) + ";" + str(hi_res_watthour_today))
if DOMOT_AC_VOLT_ID !="0" : send_to_Domoticz(str(DOMOT_AC_VOLT_ID),str(ac_volt))
if DOMOT_AC_FREQ_ID !="0" : send_to_Domoticz(str(DOMOT_AC_FREQ_ID),str(frequency))
if DOMOT_INVERTER_TEMP_ID !="0" : send_to_Domoticz(str(DOMOT_INVERTER_TEMP_ID),str(inverter_temp))
if DOMOT_VOLT_ID !="0" : send_to_Domoticz(str(DOMOT_AC_VOLT_ID),str(volt))
```

This works perfectly. Only in the last few days I often get the message "502 bad gateway" from soliscloud.
do you have experience with that?

**Comment by ZuinigeRijder** [comment]
@hotting Great that you managed to couple this to domoticz. Yes, sometimes soliscloud is overloaded, I also see this sometimes in my Raspberry Pi logging.
If I look at the logging of today, it is indeed very bad, many "502 bad gateway" messages. The python script then retries after 1 minute. Luckily I see that after some retries still the data is retrieved/send.

**Comment by PeJeWe** [comment]
I like this.
i was searching for this ;-)
I have a question:
How dus the compleet script looks like, when al the things you added are in place.
The config file i get, only the python script i don't get.
my scripting quality's are very limited.

**Comment by hotting** [comment]
@PeJeWe 
In the Python script I have inserted the code blocks I posted earlier between the next line. Starting at the bottom:
The code block is indicated between >>> and <<<

   214 send_pvoutput_data(pvoutput_string)
_**>>> The fourth block of code, calling "send_to_domoticz" <<<<**_
   215 timestamp_previous = timestamp_current


   123 return contents
   124
_**>>> The 3rd code block the routine "send_to_domoticz" <<<<**_
   125
   126 # == get_inverter_list_body ================================================= =====


   22 # == read api_secrets in soliscloud_to_pvoutput.cfg ============================
   23 parser = configparser.ConfigParser()
   24 parser.read('soliscloud_to_pvoutput.cfg')
   25 api_secrets = dict(parser.items('api_secrets'))
_**>>> The 2nd code block "read cfg file" (without the first two lines ) <<<**_
   26 # == API Secrets, fill in yours in soliscloud_to_pvoutput.cfg ================

Hopefully you have enough information with this. if you don't use pvoutput you can disable line "send_pvoutput_data(pvoutput_string)" with a #

@zuinigeRijder.
It might be an idea that I start a topic with this in the domoticz forum. I would like your permission to refer to your project. Honor where credit is due.

**Comment by ZuinigeRijder** [comment]
@hotting Maybe I will incorporate your solution in SolisCloud2PVOutput and make it configurable to send to PVOutput and/or Domoticz. 
Of course you have my permission to refer to my solution. Let me know where you are posting this in the domoticz forum, so I can follow the discussion.

**Comment by ZuinigeRijder** [comment]
@hotting Also a question about the posted code, how is frequency computed? And is it part of the config?

````
            if DOMOT_AC_FREQ_ID != "0":
                    send_to_domoticz(str(DOMOT_AC_FREQ_ID), str(frequency))
````

And I also wonder if you should have a retry implemented, like in send_pvoutput_data?

**Comment by hotting** [comment]
@ZuinigeRijder 

The content you get back from soliscloud contains more than 400 lines of information in json format. The frequency of the inverter is also passed on in this information.

   "fac": 50.01,
   "facStr": "Hz",

I had also incorporated the frequency into my solution. But that requires some more adjustments in your code. In order not to make the story of the adjustment too complicated, I had removed the freq from the earlier mention. Only the one conscious rule has been forgotten.

For completeness.
To also be able to see the freq in domoticz, you must also create a device in domoticz there. Then create a line for the frequency in the cfg file. 

         domot_ac_freq_id = 0

In the code in the section # == domoticz devices put the value of the device in a variable 

        DOMOT_AC_FREQ_ID = Domoticz['domot_ac_freq_id'].

 Then put the value of the frequency in main section in the variable "frequentie".
       
           frequentie= inverter_detail['fac']

And finally leave/add the line where your question was about to the domoticz call

         if DOMOT_AC_FREQ_ID !="0" : send_to_Domoticz(str(DOMOT_AC_FREQ_ID),str(frequentie))

Greetings HansH

**Comment by hotting** [comment]
I have created a topic on the domotcz forum

[https://www.domoticz.com/forum/viewtopic.php?t=40331](https://www.domoticz.com/forum/viewtopic.php?t=40331)

**Comment by ZuinigeRijder** [comment]
@hotting I did push my changes and released R1.9.0:
https://github.com/ZuinigeRijder/SolisCloud2PVOutput/releases/tag/R1.9.0

I hope this works as expected, I did some minor modifications to your solution (e.g. retry mechanism and some variable name renames) and left frequency out.

The configuration file has by default Domoticz disabled and PVOutput enabled.

Can you test this version for Domoticz, if it works and report back?
Thanks.

**Reply by hotting** [reply]
@ZuinigeRijder 
I'll do a test as soon as I get a chance. (I'm on vacation for a few days now)

Top job this integration

While reviewing the code, I noticed a few things
on line 267

                 if DOMOTICZ_POWER_GENERATED_ID:

Am I missing the equation, this line shouldn't be like this

                 if DOMOTICZ_POWER_GENERATED_ID != "0":

furthermore it is not better to set the variables for the domoticz devices to 0 in the cfg file so that someone is forced to give them their own value. This way the wrong device cannot be accidentally updated

                 domot_power_generated_id = 214
                 domot_ac_volt_id = 215

becomes

                 domot_power_generated_id = 0
                 domot_ac_volt_id = 0

Adjust this in the readme as well

greetings HansH

**Comment by ZuinigeRijder** [comment]
@hotting I copied this from your post earlier ;-)

````
if DOMOTICZ_POWER_GENERATED_ID:
````

Good idea to change the defaults. I will wait till you have a chance to test it, when you have time and back from vacation.

**Comment by ZuinigeRijder** [comment]
@hotting I updated your suggestions in a minor release.

**Comment by PeJeWe** [comment]
My cron is running, when I check my cron log.
Pvoutput gets updates every 5 minutes, only not domoticz. 
When I run the soliscloud_to_pvoutput.py manually and leave it open domoticz gets the updates.

**Comment by PeJeWe** [comment]
In line 277 is a faulty code
send_to_domoticz(str(DOMOTICZ_AC_VOLT_ID), str(volt))
It must be
send_to_domoticz(str(DOMOTICZ_VOLT_ID), str(volt)) without the AC part.
It otherwise won't update it to the idx domoticz.

**Comment by ZuinigeRijder** [comment]
@PeJeWe Thanks for notifying. I fixed it:
https://github.com/ZuinigeRijder/SolisCloud2PVOutput/releases/tag/R1.10.0

**Comment by hotting** [comment]
Started the script in the afternoon (including AC adjustment) and it worked fine. As planned it stopped at 11:00 PM and the next day at 5:02 AM it was started again via the cron. No problems encountered so far.

Nice that you set the default values to 0.

I do have one more request but I'll make another idea of it

Greetings HansH

**Comment by PeJeWe** [comment]
@ZuinigeRijder  can you add the daily generation?
I have currently only the actual power generation.

**Comment by ZuinigeRijder** [comment]
@PeJeWe I see that @hotting sends in the DOMOTICZ_POWER_GENERATED_ID the Watt as well as hi_res_watthour_today, separated by semicolon ;

````
                if DOMOTICZ_POWER_GENERATED_ID != 0:
                    send_to_domoticz(
                        str(DOMOTICZ_POWER_GENERATED_ID),
                        str(watt) + ";" + str(hi_res_watthour_today),
                    )
````

I think that it is indeed better to split those and just have one value send in each sensor. Something like:
````
                if DOMOTICZ_ENERGY_GENERATED_ID != 0:
                    send_to_domoticz(
                        str(DOMOTICZ_ENERGY_GENERATED_ID), str(hi_res_watthour_today)
                    )
                if DOMOTICZ_POWER_GENERATED_ID != 0:
                    send_to_domoticz(str(DOMOTICZ_POWER_GENERATED_ID), str(watt))
````

@hotting What do you think? Was there a specific reason you used one sensor with both values separated by semicolon?

**Reply by hotting** [reply]
The double values have everything to do with the created device in domoticz

In the domoticz user manual it says:


> _Electricity (instant and counter)
> /json.htm?type=command&param=udevice&idx=IDX&nvalue=0&svalue=POWER;ENERGY
> IDX = id of your device (This number can be found in the devices tab in the column "IDX")
> POWER = current power
> ENERGY = Integer, total cumulative energy in Watt-hours (Wh)_


Here you can see that two values are expected: power and energy

There is a big but.
When using this device, even more settings can be entered after the device has been created. You can choose delivery or purchase. More important is the choice between "Energy read: Computed or From device"
When using this script you have to use computed. From device expects a statement of the total energy consumption and not per day.

that is why only one device is used which displays everything

![image](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/assets/17342657/b7bcb83a-a753-4399-b60d-801bdd2741a3)


![image](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/assets/17342657/59fcd74b-5b04-4578-98a4-18819bf8482f)

**Reply by PeJeWe** [reply]
I have created the double sensor actual+counter.
I haven't seen it that two sensors were combined.
It is working fine now.

**Comment by ZuinigeRijder** [comment]
@hotting Can you help in the tweakers topic, because you are also from the Netherlands? See this post:

https://gathering.tweakers.net/forum/list_message/75349392#75349392
